vyper/venom/passes/loop_invariant_hosting.py

```diff
@@ -0,0 +1,119 @@
+from vyper.utils import OrderedSet
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.analysis.dfg import DFGAnalysis
+from vyper.venom.analysis.liveness import LivenessAnalysis
+from vyper.venom.analysis.loop_detection import NaturalLoopDetectionAnalysis
+from vyper.venom.basicblock import IRBasicBlock, IRInstruction, IRLabel, IRVariable
+from vyper.venom.function import IRFunction
+from vyper.venom.passes.base_pass import IRPass
+
+
+def _ignore_instruction(instruction: IRInstruction) -> bool:
+    return (
+        instruction.is_volatile
+        or instruction.is_bb_terminator
+        or instruction.opcode == "returndatasize"
+        or instruction.opcode == "phi"
+        or (instruction.opcode == "add" and isinstance(instruction.operands[1], IRLabel))
+        or instruction.opcode == "store"
+    )
+
+
+def _is_store(instruction: IRInstruction) -> bool:
+    return instruction.opcode == "store"
+
+
+class LoopInvariantHoisting(IRPass):
+    """
+    This pass detects invariants in loops and hoists them above the loop body.
+    Any VOLATILE_INSTRUCTIONS, BB_TERMINATORS CFG_ALTERING_INSTRUCTIONS are ignored
+    """
+
+    function: IRFunction
+    loop_analysis: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
+    dfg: DFGAnalysis
+
+    def run_pass(self):
+        self.analyses_cache.request_analysis(CFGAnalysis)
+        self.dfg = self.analyses_cache.request_analysis(DFGAnalysis)
+        loops = self.analyses_cache.request_analysis(NaturalLoopDetectionAnalysis)
+        self.loop_analysis = loops.loops
+        invalidate_dependant = False
+        while True:
+            change = False
+            for from_bb, loop in self.loop_analysis.items():
+                hoistable: list[IRInstruction] = self._get_hoistable_loop(from_bb, loop)
+                if len(hoistable) == 0:
+                    continue
+                change |= True
+                self._hoist(from_bb, hoistable)
+            if not change:
+                break
+            invalidate_dependant = True
+
+        # only need to invalidate if you did some hoisting
+        if invalidate_dependant:
+            self.analyses_cache.invalidate_analysis(LivenessAnalysis)
+
+    def _hoist(self, target_bb: IRBasicBlock, hoistable: list[IRInstruction]):
+        for inst in hoistable:
+            bb = inst.parent
+            bb.remove_instruction(inst)
+            target_bb.insert_instruction(inst, index=len(target_bb.instructions) - 1)
```

Comment:
maybe index does not need to be specified in this case

Reply:
I am modifying already terminated basic block so I it would crash on assert if the index would not be specified